File: common/src/main/java/juuxel/adorn/loot/CheckTradingStationOwnerLootFunction.java
package juuxel.adorn.loot;

import com.mojang.serialization.MapCodec;
import juuxel.adorn.block.AdornBlocks;
import juuxel.adorn.component.AdornComponentTypes;
import net.minecraft.core.component.DataComponents;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.storage.loot.LootContext;
import net.minecraft.world.level.storage.loot.functions.LootItemFunction;

public final class CheckTradingStationOwnerLootFunction implements LootItemFunction {
    public static final CheckTradingStationOwnerLootFunction INSTANCE = new CheckTradingStationOwnerLootFunction();
    public static final Builder BUILDER = new Builder();
    public static final MapCodec<CheckTradingStationOwnerLootFunction> CODEC = MapCodec.unit(INSTANCE);

    private CheckTradingStationOwnerLootFunction() {
    }

    @Override
    public ItemStack apply(ItemStack stack, LootContext lootContext) {
        if (stack.is(AdornBlocks.TRADING_STATION.get().asItem())) {
            if (!hasTrade(stack) && !hasStorage(stack)) {
                clearOwner(stack);
            }
        }

        return stack;
    }

    private boolean hasTrade(ItemStack stack) {
        var trade = stack.get(AdornComponentTypes.TRADE.get());
        return trade != null && !trade.isFullyEmpty();
    }

    private boolean hasStorage(ItemStack stack) {
        var container = stack.get(DataComponents.CONTAINER);
        if (container == null) return false;
        return container.nonEmptyItems().iterator().hasNext();
    }

    private void clearOwner(ItemStack stack) {
        stack.remove(AdornComponentTypes.TRADE_OWNER.get());
    }

    @Override
    public MapCodec<? extends LootItemFunction> codec() {
        return CODEC;
    }

    public static final class Builder implements LootItemFunction.Builder {
        private Builder() {
        }

        @Override
        public LootItemFunction build() {
            return INSTANCE;
        }
    }
}

File: common/src/main/kotlin/juuxel/adorn/loot/CheckTradingStationOwnerLootFunction.kt
package juuxel.adorn.loot

import com.mojang.serialization.Codec
import juuxel.adorn.block.AdornBlocks
import juuxel.adorn.block.entity.TradingStationBlockEntity
import juuxel.adorn.trading.Trade
import juuxel.adorn.util.InventoryComponent
import juuxel.adorn.util.getCompoundOrNull
import net.minecraft.item.BlockItem
import net.minecraft.item.ItemStack
import net.minecraft.loot.context.LootContext
import net.minecraft.loot.function.LootFunction
import net.minecraft.loot.function.LootFunctionType
import net.minecraft.nbt.NbtCompound

object CheckTradingStationOwnerLootFunction : LootFunction {
    val CODEC: Codec<CheckTradingStationOwnerLootFunction> = Codec.unit(this)

    override fun apply(stack: ItemStack, context: LootContext): ItemStack {
        if (stack.isOf(AdornBlocks.TRADING_STATION.asItem())) {
            stack.getSubNbt(BlockItem.BLOCK_ENTITY_TAG_KEY)?.let { tag ->
                if (!hasTrade(tag) && !hasStorage(tag)) {
                    clearOwner(tag)
                }
            }
        }

        return stack
    }

    private fun hasTrade(nbt: NbtCompound): Boolean {
        val tradeNbt = nbt.getCompoundOrNull(TradingStationBlockEntity.NBT_TRADE) ?: return false
        val trade = Trade.fromNbt(tradeNbt)
        return !trade.isFullyEmpty()
    }

    private fun hasStorage(nbt: NbtCompound): Boolean {
        val storageNbt = nbt.getCompoundOrNull(TradingStationBlockEntity.NBT_STORAGE) ?: return false
        val inventory = InventoryComponent(TradingStationBlockEntity.STORAGE_SIZE)
        inventory.readNbt(storageNbt)
        return !inventory.isEmpty
    }

    private fun clearOwner(nbt: NbtCompound) {
        nbt.remove(TradingStationBlockEntity.NBT_TRADING_OWNER)
        nbt.remove(TradingStationBlockEntity.NBT_TRADING_OWNER_NAME)
    }

    override fun getType(): LootFunctionType =
        AdornLootFunctionTypes.CHECK_TRADING_STATION_OWNER
}
